rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,638 @@
+    set_po_args(void **po_args); 
+}
+
+struct brgemm {
+    // Vanilla version of brgemm with no post-op or destination conversion.
+    brgemm(dim_t batch, dim_t M, dim_t N, dim_t K,
+            data_type dtA, dim_t ldA,
+            data_type dtB, dim_t ldB, 
+            data_type dtC, dim_t ldC);
+
+    // Attributes setting
+    // Those have to be set before call to generate
+
+    // If true (default), computes C = sum_i A_i * B_i
+    // If false, computes C = C + sum_i A_i * B_i
+    status_t set_add_C(bool add_C);
+
+    // adds post-operation, and conversion to final destination D.
+    status_t set_postops(post_ops &po, data_type dtD, dim_t ldD);
+
+    // scales can be applied before (upconversion) or after the GEMM operation
+    status_t set_scales(int a_scale_mask, int b_scale_mask, int d_scale_mask);
+
+    // Queries for expected layouts and temporary memory
+    pack_type get_A_pack_type() const; // Not really needed, just for consistency
+    pack_type get_B_pack_type() const;
+    size_t get_scratchpad_size() const;
+
+    // HW context handling.
+    // This currently should support AMX and SME: 
+    // - set is non-static (tilecfg for AMX or  smstart+masks setting for SME)
+    // - Release is static (tile_release for AMX or smstop for SME)
+    // - No release necessary between different calls to {re}set_hw_context
+    void set_hw_context() const;
+    void reset_hw_context() const;
+    static void release_hw_context() const;
+
+    //
+    status finalize();
+    // separate kernel generation to allow query without jit overhead
+    status generate();
+
+    // Execution function for the vanilla brgemm variant.
+    // we take base pointers to A and B, 
+    // and a vector of pairs for offsets to guarantees they are the same size
+    // The batch size is the size of the vector.
+    // pointers are void*, datatypes are specified in constructor
+    // Computes C += \sum_i A_i \cdot B_i if add_C=true
+    // C = \sum_i A_i \cdot B_i  otherwise.
+    void execute(const void *A, const void *B, 
+            const std::vector<std::pair<size_t *, size_t *>> &A_B_offsets,
+            void *C, void *scratch = nullptr);
+
+    // Execution function for the advanced brgemm<> variant
+    // Here the C matrix is just an input to accumulation
+    // final result after postop/conversion will be in D
+    // Computes D = convert(post_ops(C + \sum_i A_i \cdot B_i) if add_C=true
+    // D = convert(post_ops(\sum_i A_i \cdot B_i) otherwise
+    void execute(const void *A, const void *B, 
+            const std::vector<std::pair<size_t, size_t>> &A_B_offsets, 
+            const void *C, void *D, void *scratch = nullptr,
+            const attr_params &attr_args = attr_args());
+}
+
+struct transform {
+    // both src, dst share same dt, no fused conversion
+    transform(dim_t M, dim_t N, data_type dt, 
+            pack_type tag_src, dim_t ld_src,
+            pack_type tag_dst, dim_t ld_dst);
```

Comment:
Will there still be limitations on leading dimensions of transform (brgemm_pack_B)? 
At present, there is a limitation for ld_src https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2216e8b9c7c60d7748d668e46cd4d2594ccb26a9/src/cpu/x64/brgemm/capi/brgemm_api.cpp#L205-L208.